shap/utils/_clustering.py

```diff
@@ -228,37 +253,31 @@ def hclust(
 
     # build the distance matrix
     if metric == "xgboost_distances_r2":
-        dist_full = xgboost_distances_r2(X, y, random_state=random_state)
+        dist_full: np.ndarray = xgboost_distances_r2(X_arr, y, random_state=random_state)
 
         # build a condensed upper triangular version by taking the max distance from either direction
-        dist_list = []
-        for i in range(dist_full.shape[0]):
-            for j in range(i + 1, dist_full.shape[1]):
-                if i != j:
-                    if linkage == "single":
-                        dist_list.append(min(dist_full[i, j], dist_full[j, i]))
-                    elif linkage == "complete":
-                        dist_list.append(max(dist_full[i, j], dist_full[j, i]))
-                    elif linkage == "average":
-                        dist_list.append((dist_full[i, j] + dist_full[j, i]) / 2)
-                    else:
-                        raise ValueError("Unsupported linkage type!")
+        dist_list: list[float] = []
+        for i, j in it.combinations_with_replacement(range(len(dist_full)), 2):
+            if i == j:
+                continue
+            if linkage == "single":
+                dist_list.append(min(dist_full[i, j], dist_full[j, i]))
+            elif linkage == "complete":
+                dist_list.append(max(dist_full[i, j], dist_full[j, i]))
+            elif linkage == "average":
+                dist_list.append((dist_full[i, j] + dist_full[j, i]) / 2)
```

Comment:
Note to myself (and other maintainers):

For curiosity's sake: I did try to vectorize this operation instead of using a nested for-loop, but at least on my machine, it seems to be slower (even when `num_features = 100`) so I dropped that implementation in the end.  
Also, the bulk of the time spent is on calculating the distances (`dist_full`) anyway, so it's not really worth the effort to optimize this too much.

For reference, here's the test with the attempted "vectorized" form and the original nested for loops:
```python
def test_hclust_building_triu():
    """Tests the refactoring of building a condensed upper triangular matrix within hclust,
    when using the 'xgboost_distances_r2' metric.
    """
    pytest.importorskip("xgboost")
    from shap.utils._clustering import xgboost_distances_r2
    import time

    linkage = "single"
    X = np.random.randint(1,30, size=(500, 50))
    y = np.where(X[:, 0] > 10, 1, 0)

    dist_full = xgboost_distances_r2(X, y, random_state=64)

    t0 = time.monotonic()
    # New logic (proposed)
    if linkage == "single":
        dist_new = np.minimum(dist_full, dist_full.T)
    elif linkage == "complete":
        dist_new = np.maximum(dist_full, dist_full.T)
    elif linkage == "average":
        dist_new = (dist_full + dist_full.T) / 2
    else:
        raise ValueError("Unsupported linkage type!")
    dist_new = dist_new[np.triu_indices_from(dist_new, 1)]
    t1 = time.monotonic()
    print(f">>> New {t1 - t0}")

    # Old logic (manual)
    dist = []
    for i in range(dist_full.shape[0]):
        for j in range(i + 1, dist_full.shape[1]):
            if i != j:
                if linkage == "single":
                    dist.append(min(dist_full[i, j], dist_full[j, i]))
                elif linkage == "complete":
                    dist.append(max(dist_full[i, j], dist_full[j, i]))
                elif linkage == "average":
                    dist.append((dist_full[i, j] + dist_full[j, i]) / 2)
                else:
                    raise ValueError("Unsupported linkage type!")
    dist_old = np.array(dist)
    t2 = time.monotonic()
    print(f">>> Old {t2 - t1}")

    assert (dist_new == dist_old).all()
```

And the results when running via pytest:

```
$ pytest -r P tests/utils/test_clustering.py::test_hclust_building_triu
...
_________________________________ test_hclust_building_triu _________________________________
--------------------------------------- Captured stdout call ----------------------------------------
>>> New 0.0005963749717921019
>>> Old 0.00045070797204971313
```

Reply:
Interesting, did you test where the break even happens (I strongly assume that numpy gets way faster for larger matrices)?

Reply:
@CloseChoice Not really, at X.shape=(500,100) (i.e., 500 samples, 100 feature dimensions), the xgboost `dist_full` was already taking 40s to run on my machine (IIRC, it's a non-trivial amount of time, let's just put it that way).  
Whereas building the linkages was performing at most 100**2 operations and finishing in <0.01s.  
The numpy vectorized way is still slower in this case.

I have the same expectations as you (that the numpy version should be faster, but I think it's unnecessary to test further :x).
In most realistic cases, we have around this order of magnitude of features anyway.